File: src/actions.mjs
// Interaction = link/form graph traversal (SPEC §6). No JS runs, so we resolve
// the page's *intent graph* rather than firing synthetic events:
//   - <a href>            → navigate to the resolved URL
//   - <form action method> → serialize successful controls, navigate (GET=query, POST=body)
//   - everything else      → inert in Lane A (flagged jsHandler upstream)
//
// These are pure helpers over turbo-dom nodes; Page wires them to its navigation.

import { resolve } from "./url.mjs";

const SUBMIT_TYPES = new Set(["submit", "button", "reset", "image"]);

/** Set a control's live value (and checked state for checkbox/radio). */
export function fillValue(el, value) {
  const tag = el.tagName.toLowerCase();
  const type = el.getAttribute("type")?.toLowerCase();
  if (tag === "input" && (type === "checkbox" || type === "radio")) {
    el.checked = Boolean(value);
    return;
  }
  el.value = value == null ? "" : String(value);
}

// Per-control current value: prefer the live `.value` property (reflects fill()),
// fall back to the attribute for elements whose property turbo-dom leaves unset.
function controlValue(el) {
  const v = el.value;
  if (v !== undefined && v !== null) return String(v);
  return el.getAttribute("value") ?? "";
}

// Selected <option> [name,value] pairs for a <select> control.
function selectPairs(el, name) {
  const out = [];
  const opts = el.querySelectorAll("option");
  for (let j = 0; j < opts.length; j++) {
    if (opts[j].selected) {
      out.push([name, opts[j].getAttribute("value") ?? opts[j].textContent.trim()]);
    }
  }
  return out;
}

function isSubmitControl(tag, type) {
  return tag === "button" || (tag === "input" && SUBMIT_TYPES.has(type));
}

function isCheckable(tag, type) {
  return tag === "input" && (type === "checkbox" || type === "radio");
}

// [name,value] pairs for a typed (non-submit/non-select) control.
function typedPairs(el, name, tag, type) {
  if (isCheckable(tag, type)) {
    return el.checked ? [[name, el.getAttribute("value") ?? "on"]] : [];
  }
  if (tag === "select") return selectPairs(el, name);
  return [[name, controlValue(el)]];
}

// Classify one control into its successful [name,value] pairs (possibly none).
// Mirrors the HTML successful-controls rules; `submitter` is the activated button.
function controlPairs(el, submitter) {
  const name = el.getAttribute("name");
  if (!name || el.getAttribute("disabled") !== null) return [];

  const tag = el.tagName.toLowerCase();
  const type = el.getAttribute("type")?.toLowerCase();

  if (isSubmitControl(tag, type)) {
    // Only the activated submitter is successful.
    return el === submitter ? [[name, controlValue(el)]] : [];
  }
  return typedPairs(el, name, tag, type);
}

/**
 * Collect a form's *successful controls* (HTML form-submission subset) as
 * [name, value] pairs. `submitter` (the activated submit button), if given and
 * named, contributes its name/value.
 */
export function serializeForm(form, submitter) {
  const pairs = [];
  const controls = form.elements ?? form.querySelectorAll("input,select,textarea,button");
  for (let i = 0; i < controls.length; i++) {
    for (const pair of controlPairs(controls[i], submitter)) pairs.push(pair);
  }
  return pairs;
}

/**
 * Build the navigation a form submit produces.
 * @returns {{ method:"GET"|"POST", url:string, body?:string, contentType?:string }}
 */
function buildGet(actionUrl, pairs) {
  const u = new URL(actionUrl);
  u.search = new URLSearchParams(pairs).toString();
  return { method: "GET", url: u.href };
}

function buildPost(actionUrl, pairs) {
  return {
    method: "POST",
    url: actionUrl,
    body: new URLSearchParams(pairs).toString(),
    contentType: "application/x-www-form-urlencoded",
  };
}

let boundarySeq = 0;
function makeBoundary() {
  boundarySeq += 1;
  return `----turboCrawlFormBoundary${boundarySeq}`;
}

function multipartPart(boundary, name, value) {
  return `--${boundary}\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${value}\r\n`;
}

// multipart/form-data body for forms declaring enctype=multipart (SPEC §6).
// Text controls only — file inputs can't be read without JS in Lane A.
function buildMultipart(actionUrl, pairs) {
  const boundary = makeBoundary();
  let body = "";
  for (const [name, value] of pairs) body += multipartPart(boundary, name, value);
  body += `--${boundary}--\r\n`;
  return {
    method: "POST",
    url: actionUrl,
    body,
    contentType: `multipart/form-data; boundary=${boundary}`,
  };
}

function formMethod(form) {
  return (form.getAttribute("method") || "GET").toUpperCase() === "POST" ? "POST" : "GET";
}

function isMultipart(form) {
  return (form.getAttribute("enctype") || "").toLowerCase().includes("multipart");
}

function formActionUrl(form, baseUrl) {
  return resolve(baseUrl, form.getAttribute("action") ?? "") ?? baseUrl;
}

export function buildSubmission(form, baseUrl, submitter) {
  const actionUrl = formActionUrl(form, baseUrl);
  const pairs = serializeForm(form, submitter);
  if (formMethod(form) === "GET") return buildGet(actionUrl, pairs);
  return isMultipart(form) ? buildMultipart(actionUrl, pairs) : buildPost(actionUrl, pairs);
}

File: rust/crates/turbo-surf-view/src/actions.rs
//! Interaction = link/form graph traversal (port of `src/actions.mjs`, SPEC §6).
//! No JS runs, so we resolve the page's *intent graph*: `<a href>` navigates,
//! `<form>` serializes its successful controls into a GET/POST submission.
//! Pure over the Tree (Page wires these to navigation). Control "values" are the
//! `value`/`checked`/`selected` attributes (see `dom_ops` for the mutators).

use crate::dom_ops::text_of;
use std::sync::atomic::{AtomicUsize, Ordering};
use turbo_dom_parser::rtdom::{DocumentExt, Tree};
use turbo_surf_core::url::resolve;
use url::form_urlencoded;

const SUBMIT_TYPES: &[&str] = &["submit", "button", "reset", "image"];

/// What clicking an element does in Lane A (no JS).
#[derive(Debug, Clone, PartialEq)]
pub enum ClickIntent {
    /// Follow an `<a href>` to this absolute URL.
    Navigate(String),
    /// Submit the owning `<form>`.
    Submit(Submission),
    /// A JS-only handler — nothing to fire without JS.
    Inert,
}

/// A navigation produced by submitting a form.
#[derive(Debug, Clone, PartialEq)]
pub struct Submission {
    pub method: String,
    pub url: String,
    pub body: Option<String>,
    pub content_type: Option<String>,
}

/// Set a control's value (and checked state for checkbox/radio) via attributes.
pub fn fill_value(tree: &mut Tree, h: u32, value: &str) {
    let tag = tree.tag_name(h).unwrap_or_default();
    let ty = tree
        .get_attribute(h, "type")
        .map(|t| t.to_ascii_lowercase());
    if tag == "INPUT" && matches!(ty.as_deref(), Some("checkbox") | Some("radio")) {
        set_bool_attr(tree, h, "checked", !value.is_empty());
        return;
    }
    tree.set_attribute(h, "value", value);
}

fn set_bool_attr(tree: &mut Tree, h: u32, attr: &str, on: bool) {
    if on {
        tree.set_attribute(h, attr, "");
    } else {
        tree.remove_attribute(h, attr);
    }
}

fn lower_tag(tree: &Tree, h: u32) -> String {
    tree.tag_name(h).unwrap_or_default().to_ascii_lowercase()
}

fn lower_type(tree: &Tree, h: u32) -> Option<String> {
    tree.get_attribute(h, "type")
        .map(|t| t.to_ascii_lowercase())
}

fn control_value(tree: &Tree, h: u32) -> String {
    tree.get_attribute(h, "value").unwrap_or("").to_string()
}

fn is_submit_control(tag: &str, ty: Option<&str>) -> bool {
    tag == "button" || (tag == "input" && ty.is_some_and(|t| SUBMIT_TYPES.contains(&t)))
}

fn is_checkable(tag: &str, ty: Option<&str>) -> bool {
    tag == "input" && matches!(ty, Some("checkbox") | Some("radio"))
}

fn select_pairs(tree: &Tree, select: u32, name: &str) -> Vec<(String, String)> {
    tree.node(select)
        .query_selector_all("option")
        .iter()
        .filter(|o| tree.get_attribute(o.handle(), "selected").is_some())
        .map(|o| {
            let v = tree
                .get_attribute(o.handle(), "value")
                .map_or_else(|| text_of(tree, o.handle()), str::to_string);
            (name.to_string(), v)
        })
        .collect()
}

fn typed_pairs(
    tree: &Tree,
    h: u32,
    name: &str,
    tag: &str,
    ty: Option<&str>,
) -> Vec<(String, String)> {
    if is_checkable(tag, ty) {
        return match tree.get_attribute(h, "checked") {
            Some(_) => vec![(
                name.to_string(),
                tree.get_attribute(h, "value").unwrap_or("on").to_string(),
            )],
            None => Vec::new(),
        };
    }
    if tag == "select" {
        return select_pairs(tree, h, name);
    }
    vec![(name.to_string(), control_value(tree, h))]
}

// One control → its successful [name,value] pairs (HTML successful-controls).
fn control_pairs(tree: &Tree, h: u32, submitter: Option<u32>) -> Vec<(String, String)> {
    let Some(name) = tree.get_attribute(h, "name") else {
        return Vec::new();
    };
    if name.is_empty() || tree.get_attribute(h, "disabled").is_some() {
        return Vec::new();
    }
    let tag = lower_tag(tree, h);
    let ty = lower_type(tree, h);
    if is_submit_control(&tag, ty.as_deref()) {
        // Only the activated submitter is successful.
        return if Some(h) == submitter {
            vec![(name.to_string(), control_value(tree, h))]
        } else {
            Vec::new()
        };
    }
    typed_pairs(tree, h, name, &tag, ty.as_deref())
}

/// A form's successful controls as `[name, value]` pairs. `submitter` (the
/// activated submit button), if named, contributes its name/value.
pub fn serialize_form(tree: &Tree, form: u32, submitter: Option<u32>) -> Vec<(String, String)> {
    let controls: Vec<u32> = tree
        .node(form)
        .query_selector_all("input,select,textarea,button")
        .iter()
        .map(|c| c.handle())
        .collect();
    controls
        .into_iter()
        .flat_map(|c| control_pairs(tree, c, submitter))
        .collect()
}

fn encode(pairs: &[(String, String)]) -> String {
    let mut ser = form_urlencoded::Serializer::new(String::new());
    for (k, v) in pairs {
        ser.append_pair(k, v);
    }
    ser.finish()
}

fn build_get(action: &str, pairs: &[(String, String)]) -> Submission {
    let url = match url::Url::parse(action) {
        Ok(mut u) => {
            u.set_query(Some(&encode(pairs)));
            u.to_string()
        }
        Err(_) => action.to_string(), // defensive: action is pre-resolved to absolute
    };
    Submission {
        method: "GET".into(),
        url,
        body: None,
        content_type: None,
    }
}

fn build_post(action: &str, pairs: &[(String, String)]) -> Submission {
    Submission {
        method: "POST".into(),
        url: action.to_string(),
        body: Some(encode(pairs)),
        content_type: Some("application/x-www-form-urlencoded".into()),
    }
}

static BOUNDARY_SEQ: AtomicUsize = AtomicUsize::new(0);

fn build_multipart(action: &str, pairs: &[(String, String)]) -> Submission {
    let seq = BOUNDARY_SEQ.fetch_add(1, Ordering::Relaxed) + 1;
    let boundary = format!("----turboCrawlFormBoundary{seq}");
    let mut body = String::new();
    for (name, value) in pairs {
        body.push_str(&format!(
            "--{boundary}\r\nContent-Disposition: form-data; name=\"{name}\"\r\n\r\n{value}\r\n"
        ));
    }
    body.push_str(&format!("--{boundary}--\r\n"));
    Submission {
        method: "POST".into(),
        url: action.to_string(),
        body: Some(body),
        content_type: Some(format!("multipart/form-data; boundary={boundary}")),
    }
}

fn form_method(tree: &Tree, form: u32) -> &'static str {
    let m = tree
        .get_attribute(form, "method")
        .unwrap_or("GET")
        .to_ascii_uppercase();
    if m == "POST" {
        "POST"
    } else {
        "GET"
    }
}

fn is_multipart(tree: &Tree, form: u32) -> bool {
    tree.get_attribute(form, "enctype")
        .unwrap_or("")
        .to_ascii_lowercase()
        .contains("multipart")
}

fn form_action_url(tree: &Tree, form: u32, base: &str) -> String {
    let action = tree.get_attribute(form, "action").unwrap_or("");
    resolve(base, action).unwrap_or_else(|| base.to_string())
}

fn ancestor_anchor_href(tree: &Tree, h: u32) -> Option<String> {
    let mut cur = Some(h);
    while let Some(n) = cur {
        if tree.tag_name(n).as_deref() == Some("A") {
            if let Some(href) = tree.get_attribute(n, "href") {
                return Some(href.to_string());
            }
        }
        cur = tree.parent(n);
    }
    None
}

fn ancestor_form(tree: &Tree, h: u32) -> Option<u32> {
    let mut cur = Some(h);
    while let Some(n) = cur {
        if tree.tag_name(n).as_deref() == Some("FORM") {
            return Some(n);
        }
        cur = tree.parent(n);
    }
    None
}

fn is_submitter(tree: &Tree, h: u32) -> bool {
    let tag = tree.tag_name(h).unwrap_or_default();
    let ty = tree
        .get_attribute(h, "type")
        .map(|t| t.to_ascii_lowercase());
    tag == "BUTTON" || (tag == "INPUT" && matches!(ty.as_deref(), Some("submit") | Some("image")))
}

/// Resolve what clicking element `h` does (no JS): follow an `<a href>`, submit
/// the owning `<form>`, or inert.
pub fn click_intent(tree: &Tree, h: u32, base: &str) -> ClickIntent {
    if let Some(href) = ancestor_anchor_href(tree, h) {
        return ClickIntent::Navigate(resolve(base, &href).unwrap_or(href));
    }
    if let Some(form) = ancestor_form(tree, h) {
        let submitter = is_submitter(tree, h).then_some(h);
        return ClickIntent::Submit(build_submission(tree, form, base, submitter));
    }
    ClickIntent::Inert
}

/// Build the navigation a form submit produces.
pub fn build_submission(tree: &Tree, form: u32, base: &str, submitter: Option<u32>) -> Submission {
    let action = form_action_url(tree, form, base);
    let pairs = serialize_form(tree, form, submitter);
    if form_method(tree, form) == "GET" {
        return build_get(&action, &pairs);
    }
    if is_multipart(tree, form) {
        build_multipart(&action, &pairs)
    } else {
        build_post(&action, &pairs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn first(tree: &Tree, sel: &str) -> u32 {
        tree.query_selector(sel).unwrap()
    }

    const BASE: &str = "https://x.test/page";

    #[test]
    fn fill_text_and_checkbox() {
        let mut tree = Tree::parse("<input id='t'><input id='c' type='checkbox'>");
        let (t, c) = (first(&tree, "#t"), first(&tree, "#c"));
        fill_value(&mut tree, t, "hello");
        assert_eq!(tree.get_attribute(t, "value"), Some("hello"));
        fill_value(&mut tree, c, "on");
        assert!(tree.get_attribute(c, "checked").is_some());
        fill_value(&mut tree, c, "");
        assert!(tree.get_attribute(c, "checked").is_none());
    }

    #[test]
    fn get_form_serializes_to_query() {
        let tree = Tree::parse(
            "<form action='/search' method='get'><input name='q' value='rust'><input name='p' value='2'></form>",
        );
        let s = build_submission(&tree, first(&tree, "form"), BASE, None);
        assert_eq!(s.method, "GET");
        assert_eq!(s.url, "https://x.test/search?q=rust&p=2");
        assert_eq!(s.body, None);
    }

    #[test]
    fn post_form_serializes_to_body() {
        let tree =
            Tree::parse("<form action='/submit' method='POST'><input name='a' value='1'></form>");
        let s = build_submission(&tree, first(&tree, "form"), BASE, None);
        assert_eq!(s.method, "POST");
        assert_eq!(s.url, "https://x.test/submit");
        assert_eq!(s.body.as_deref(), Some("a=1"));
        assert_eq!(
            s.content_type.as_deref(),
            Some("application/x-www-form-urlencoded")
        );
    }

    #[test]
    fn disabled_and_unnamed_controls_excluded() {
        let tree = Tree::parse(
            "<form method='get'><input name='a' value='1'><input name='b' value='2' disabled><input value='3'></form>",
        );
        let s = build_submission(&tree, first(&tree, "form"), BASE, None);
        assert_eq!(s.url, "https://x.test/page?a=1");
    }

    #[test]
    fn only_activated_submitter_contributes() {
        let tree = Tree::parse(
            "<form method='get'><input name='q' value='x'><button name='go' value='1'>Go</button><button name='cancel' value='1'>Cancel</button></form>",
        );
        let go = first(&tree, "[name=go]");
        let s = build_submission(&tree, first(&tree, "form"), BASE, Some(go));
        assert_eq!(s.url, "https://x.test/page?q=x&go=1");
    }

    #[test]
    fn checkbox_and_select_serialization() {
        let tree = Tree::parse(
            "<form method='get'><input type='checkbox' name='c' checked><input type='checkbox' name='d'><select name='s'><option value='a'>A</option><option value='b' selected>B</option></select></form>",
        );
        let s = build_submission(&tree, first(&tree, "form"), BASE, None);
        // checked checkbox → c=on; unchecked d omitted; selected option → s=b
        assert_eq!(s.url, "https://x.test/page?c=on&s=b");
    }

    #[test]
    fn multipart_form_body() {
        let tree = Tree::parse(
            "<form method='post' enctype='multipart/form-data'><input name='a' value='1'></form>",
        );
        let s = build_submission(&tree, first(&tree, "form"), BASE, None);
        assert_eq!(s.method, "POST");
        assert!(s
            .content_type
            .as_deref()
            .unwrap()
            .starts_with("multipart/form-data; boundary="));
        let body = s.body.unwrap();
        assert!(body.contains("Content-Disposition: form-data; name=\"a\""));
        assert!(body.contains("\r\n\r\n1\r\n"));
        assert!(body.trim_end().ends_with("--"));
    }

    #[test]
    fn action_defaults_to_base_when_absent() {
        let tree = Tree::parse("<form method='get'><input name='a' value='1'></form>");
        let s = build_submission(&tree, first(&tree, "form"), BASE, None);
        assert_eq!(s.url, "https://x.test/page?a=1");
    }
}
